### DISCUSSION TITLE
Can I override the redirect to /api/auth/callback/ when using Rest?

### DISCUSSION THREAD

**Discussion by leo-petrucci** [nextauthjs/next-auth#4319] [Help] [answered]
I need to use Next Auth's rest endpoints to authenticate. From asking in previous thread I was able to get Next Auth to return the authorization URL for my provider, but I'm having trouble figuring out if I can customise the `redirect_uri` parameter when using the `/api/auth/signin/:provider` endpoint.

```
    await fetch(
      "https://mysite.com/api/auth/signin/twitter",
      {
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
        },
        body: new URLSearchParams({
          csrfToken,
          json: "true",
          callbackUrl: "http://someothersite.com",
        }),
        credentials: "include",
      }
    )
```

`callbackUrl` in this case is where Next Auth will redirect my user **AFTER** Twitter redirects me to `/api/auth/callback/:provider`. 

What If I wanted to redirect somewhere else, and then POST and GET to `/api/auth/callback/:provider`? Is that possible?

### Edit
Looking at the source it looks like `redirect_uri` is set by [this `options` object](https://github.com/nextauthjs/next-auth/blob/main/packages/next-auth/src/core/routes/signin.ts#L26). I'm unclear how this is passed to the `signin` function however.

**Comment by renevangsgaardjp** [comment]
I customize the `redirect_uri` when performing the sign in. 
```
signIn("keycloak", undefined, {
  // Make sure to overwrite the redirect_uri.
  redirect_uri: `https://${host}/api/auth/callback/keycloak`,
})
```
I have troubles though, when next-auth further on tries to fetch the id token. See more here https://github.com/nextauthjs/next-auth/issues/4668.

**Reply by leo-petrucci** [reply]
That's really really strange. It looks like the `signIn` function respects `redirect_url`, but doing it via `fetch` does not work.

**Comment by leo-petrucci** [comment, marked as answer]
I figured out how to use a custom `redirect_uri` on signin. The value isn't to be passed to the body, but rather appended as a query to the url.

```
  const doLogin = async () => {
    const { csrfToken }: { csrfToken: string } = await fetch(
      "http://localhost:3000/api/auth/csrf",
      {
        credentials: "include",
      }
    ).then((res) => res.json());

    const { url } = await fetch(
      `http://localhost:3000/api/auth/signin/twitter?${new URLSearchParams({
        redirect_uri: "http://localhost:3000/custom-callback",
      })}`,
      {
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
        },
        body: new URLSearchParams({
          csrfToken,
          json: "true",
        }),
        credentials: "include",
      }
    ).then((res) => res.json());

    router.push(url);
  };
```
